src/commonMain/kotlin/io/ktor/io/Buffer.kt

```diff
@@ -1,43 +1,289 @@
+     *
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public abstract fun loadIntAt(index: Int): Int
 
-    public abstract fun release()
-    public abstract fun compact()
+    /**
+     * Reads [Long] at specific [index].
+     *
+     * The operation doesn't modify [readIndex] or [writeIndex].
+     *
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public abstract fun loadLongAt(index: Int): Long
 
-    public companion object {
-        public val Empty: Buffer = ByteArrayBuffer(0)
+    /**
+     * Writes [Boolean] at specific [index].
+     *
+     * The operation doesn't modify [readIndex] or [writeIndex].
+     *
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public open fun storeBooleanAt(index: Int, value: Boolean) {
+        storeByteAt(index, if (value) 1.toByte() else 0.toByte())
+    }
+
+    /**
+     * Writes [Byte] at specific [index].
+     *
+     * The operation doesn't modify [readIndex] or [writeIndex].
+     *
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public abstract fun storeByteAt(index: Int, value: Byte)
+
+    /**
+     * Writes [Short] at specific [index].
+     *
+     * The operation doesn't modify [readIndex] or [writeIndex].
+     *
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public abstract fun storeShortAt(index: Int, value: Short)
+
+    /**
+     * Writes [Int] at specific [index].
+     *
+     * The operation doesn't modify [readIndex] or [writeIndex].
+     *
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public abstract fun storeIntAt(index: Int, value: Int)
+
+    /**
+     * Writes [Long] at specific [index].
+     *
+     * The operation doesn't modify [readIndex] or [writeIndex].
+     *
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public abstract fun storeLongAt(index: Int, value: Long)
+
+    /**
+     * Writes as many bytes as possible from the [buffer] at specific [index].
+     *
+     * The operation modifies [readIndex] in [buffer], but doesn't modify [writeIndex] in this buffer.
+     *
+     * @return Number of written bytes: `min(availableForWrite, buffer.availableForRead)`
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public abstract fun storeBufferAt(index: Int, value: Buffer): Int
+
+    /**
+     * Copy all bytes from [value] between [startPosition] and [endPosition] to the buffer at specific [index].
+     *
+     * The operation doesn't modify [readIndex] or [writeIndex].
+     *
+     * @return Number of written bytes: `min(availableForWrite, endPosition - startPosition)`
+     * @throws IndexOutOfBoundsException if [index] is greater than [capacity].
+     */
+    public abstract fun storeArrayAt(
```

Comment:
storeByteArrayAt? As I think someone may want to have other extensions, like `storeIntArrayAt` and so on

Reply:
not sure, what you want to make an extension? :)

Reply:
about array of primitive types like `IntArray`

Reply:
from first sight it looks ok, but
on JVM there is not only ByteBuffer, but also f.e. IntBuffer, which has fast copy of IntArrays, which should work faster, than just reading Ints one by one
Not sure still, if we want to support such things or not, but if so, all those methods should be `open fun` inside `Buffer` class with default implementation

Reply:
I'm going to add default implementation for all methods in the Buffer except `getByteAt()`